### PR TITLE
Added support for config files

### PR DESCRIPTION
**Pull request by davvil**
**Update:**

As per comments, I updated the format to YAML. The corresponding updated example for the config file is now:

```
source: /Users/dvilar/corpora/iwslt/tokenize/de-en/train.tags.de-en.de
target: /Users/dvilar/corpora/iwslt/tokenize/de-en/train.tags.de-en.en
validation_source: /Users/dvilar/corpora/iwslt/tokenize/de-en/IWSLT16.TED.dev2010.de-en.de.txt
validation_target: /Users/dvilar/corpora/iwslt/tokenize/de-en/IWSLT16.TED.dev2010.de-en.en.txt
output: out
use_cpu: true
```

Fun fact: the original json config file is also correctly parse as yaml :-)

**Original description:**

Specify a config file with --config. Command line parameters have precedence over the values read from the config file, as expected.

The config file is a json serialization of the namespace returned by argparse, casted to a dictionary. The file args.json produced by the training can be used as config file.

The config file does not need to be complete. Missing parameters will be read from the command line or will take default values.

Minimal working example (`python -m sockeye.train --config config.json` starts a training, cpu activated):
```
{
  "source": "/Users/dvilar/corpora/iwslt/tokenize/de-en/train.tags.de-en.de",
  "target": "/Users/dvilar/corpora/iwslt/tokenize/de-en/train.tags.de-en.en",
  "validation_source": "/Users/dvilar/corpora/iwslt/tokenize/de-en/IWSLT16.TED.dev2010.de-en.de.txt",
  "validation_target": "/Users/dvilar/corpora/iwslt/tokenize/de-en/IWSLT16.TED.dev2010.de-en.en.txt",
  "output": "out",
  "use_cpu": true
}
```

Additionally the functionality introduced in 5a3bf5f8 has been reimplemented, accessing the argument_definitions member of ConfigArgumentParser.

### Note about the PR ###
This PR includes two commits. The first one mainly reverts arguments.py to pre-5a3bf5f8. The second one is the actual "meat" of this PR. Looking at the commits separately probably will provide a better overview.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/arguments.py [hunk 1]
```diff
@@ -15,14 +15,58 @@
...
+import types
 from typing import Callable, Optional
 
 from sockeye.lr_scheduler import LearningRateSchedulerFixedStep
 from . import constants as C
 from . import data_io
 
+# Adapted from https://stackoverflow.com/questions/28579661/getting-required-option-from-namespace-in-python
+class ConfigArgumentParser(argparse.ArgumentParser):
+    def __init__(self, *args, **kwargs):
```
Comment (fhieber) [inline sockeye/arguments.py]: add '-> None'

File: sockeye/arguments.py [hunk 2]
```diff
@@ -15,14 +15,58 @@
...
+        super().__init__(*args, **kwargs)
+        self.argument_definitions = {}
+        self.argument_actions = []
+        self._overwrite_add_argument(self)
+        self.add_argument("--config", help="Config file in JSON format.", type=str)  # Not FileType so that we can get the path here
```
Comment (fhieber) [inline sockeye/arguments.py]: overly long line I think

File: sockeye/arguments.py [hunk 3]
```diff
@@ -15,14 +15,58 @@
...
+
+    def register_argument(self, _action, *args, **kwargs):
```
Comment (fhieber) [inline sockeye/arguments.py]: will arguments also be registered if there is no group?

File: sockeye/arguments.py [hunk 4]
```diff
@@ -15,14 +15,58 @@
...
 import sys
+import types
 from typing import Callable, Optional
 
 from sockeye.lr_scheduler import LearningRateSchedulerFixedStep
 from . import constants as C
 from . import data_io
 
+# Adapted from https://stackoverflow.com/questions/28579661/getting-required-option-from-namespace-in-python
+class ConfigArgumentParser(argparse.ArgumentParser):
```
Comment (fhieber) [inline sockeye/arguments.py]: This could use a docstring explaining what it does and why :)

**Comment by fhieber** [comment]
This deserves a new minor version and a changelog entry as it is potentially interesting to many users.

**Comment by fhieber** [comment]
@yx3zhang: FYI

**Comment by yx3zhang** [comment]
Just a quick comment, this might be solving the 10% on top of the 90% this PR already covers. It would be more user friendly to support .yaml so documentation of each parameter can be provided in line.

**Comment by fhieber** [comment]
Good point, we already depend on `pyyaml`, so switching from JSON to YAML shouldn't be difficult. We have been dumping the arguments to `<model>/args.json` only for logging purposes for a while, but I don't think users rely on it at this point.
So switching to YAML with a minor version bump should be fine.

**Comment by fhieber** [comment]
@davvil #403 is a PR on top of yours to change to YAML

**Comment by davvil** [comment]
@fhieber hadn't seen your commit, I implemented it myself independently.

As you pointed out, I originally chose json because it was the format we were dumping the args originally. I also found yaml more convenient.

File: sockeye/arguments.py (hunk 3 above)
Comment (davvil) [reply sockeye/arguments.py]: Yes they will. Line 34 calls _overwrite_add_argument on itself for supporting the non-group case. What is not supported (but not used in sockeye) are sub-parsers [and possibly other exotic argparse use cases].

**Comment by davvil** [comment]
Sure, go ahead!

**Pull request merged**